### ISSUE TITLE
second order differential of fused_batch_norm makes a large number of nodes.

### ISSUE THREAD

**Issue by laket** [tensorflow/tensorflow#17856]
I found that the second order differential of tf.nn.fused_batch_norm has much larger number of nodes than that of tf.nn.batch_normalization. 

I wrote a sample program which has 7 batch normalization layers. Using tf.nn.fused_batch_norm, it has about 22,000 nodes. On the other hand, using tf.nn.batch_normalization, it has only 1700 nodes (I counted with tensorboard). Is this a reasonable behavior?

### Source Code
```
import numpy as np
import tensorflow as tf

use_fused_batch_norm = False

with tf.name_scope("network"):
    featmap = tf.constant(np.random.rand(3,3,32,32), dtype=tf.float32)

    for i in range(7):
        with tf.name_scope("layer{}".format(i)):
            if use_fused_batch_norm:
                beta = tf.Variable(np.zeros(3), name="beta", dtype=np.float32)
                gamma = tf.Variable(np.ones(3), name="gamma", dtype=np.float32)
                featmap, _, _ = tf.nn.fused_batch_norm(featmap, gamma, beta, data_format="NCHW")

            else:
                beta = tf.Variable(np.zeros([1,3,1,1]), name="beta", dtype=np.float32)
                gamma = tf.Variable(np.ones([1,3,1,1]), name="gamma", dtype=np.float32)
                mu, sigma = tf.nn.moments(featmap, axes=[0,2,3], keep_dims=True)

                featmap = tf.nn.batch_normalization(featmap, mu, sigma, beta, gamma, variance_epsilon=1e-07)

    y = tf.reduce_mean(featmap, axis=[1,2,3])

W = tf.trainable_variables()

with tf.name_scope("rop"):
    temporary = tf.ones_like(y)
    grad      = tf.gradients(y, W, grad_ys=temporary)
    r         = [tf.ones_like(t) for t in W]
    ggrad     = tf.gradients(grad, temporary, grad_ys=r)[0]


with tf.Session() as sess:
    summary_writer = tf.summary.FileWriter("log",
                                           sess.graph)

```

------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Ubuntu 16.04
- **TensorFlow installed from (source or binary)**:
binary
- **TensorFlow version (use command below)**:
('v1.5.0-0-g37aa430d84', '1.5.0')
- **Python version**: 
2.7.12
- **CUDA/cuDNN version**:
cuda9 / cudnn 7.0.5
- **GPU model and memory**:
GTX 1080, 8GB
- **Exact command to reproduce**:
python main.py
- **Bazel version** 
N/A

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Bazel version

**Comment by laket** [comment]
FusedBatchNormGradGrad uses gradients funcion to calculate gradients in  [nn_grad.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/nn_grad.py#L936).

```
  grad_grad_y, grad_x, grad_scale = gradients_impl.gradients(
      [grad_x, grad_scale, grad_offset], [grad_y, x, scale], grad_initial)
```

Then derivative of (grad_y) with respect to x is calculated. I think this makes a large number of nodes of the above network. The network size become much smaller, when we use stop_gradients in the above snippet like

```
  grad_grad_y, grad_x, grad_scale = gradients_impl.gradients(
      [grad_x, grad_scale, grad_offset], [grad_y, x, scale], grad_initial, stop_gradients=[grad_y, x, scale])
```

Can we use stop_gradients?

**Comment by cy89** [comment]
@ebrevdo @andydavis1 could you comment about whether this is WAI?

**Comment by ebrevdo** [comment]
This is fine. Grappler prunes many of them at runtime.

**Comment by laket** [comment]
@ebrevdo thank you for reply. 

Do you mean that the graph is shrinked at sess.run, so the temporary big graph is not important?

When I use second order differential of fused_batch_norm in a big graph like Resnet, memory is exhausted at tf.gradients.
So my program can't reach sess.run. 

If we can optimize graph at graph construction time, I think we should do it.

**Comment by ebrevdo** [comment]
The graph is rewritten at session.run, yes. That doesn't help you though.
Thousands of ops seems a bit excessive.  Can you identify which forward op
exactly causes the gradient graph to blow up?

**Comment by laket** [comment]
As described in my second [post](https://github.com/tensorflow/tensorflow/issues/17856#issuecomment-375599868) , 
process in FusedBatchNormGradGrad cause the gradient graph to blow up.

When I use stop_gradient in the function, the graph is much shrinked.
